### ISSUE TITLE
Working ScipyParticle but error in JITParticle mode

### ISSUE THREAD

**Issue by ursho2552** [OceanParcels/parcels#692]
Hi, recently started working with parcels for one of my PhD projects, as the ability to define movement patterns for individual particles seems extremely interesting, and it seems to be a user friendly alternative to other Lagrangian tools. However, recently I stumbled upon some difficulties when simulating particles in the JITParticle mode. Until now, I was able to simulate particles in Scipy mode, but when changing to JIT it throws the error `KernelError: 0`.  Tried simulating one particle at a time, and it sometimes worked, but for some initial positions it didn't. This although the simulation in Scipy mode always worked.

Maybe you have encountered this issue already, or maybe it is a problem on my end with the compiler.

My ROMS setup uses a telescopic, curvilinear grid with an s-coordinate system. Similar to the setup in issue  [#502](https://github.com/OceanParcels/parcels/issues/502) I calculated the z levels and assumed they are constant in time.

My netcdf structure for the velocities: 
```
netcdf nco_time_uvw_pactcs30_1979 {
dimensions:
	eta_rho = 518 ;
	xi_rho = 604 ;
	time = UNLIMITED ; // (365 currently)
	s_rho = 64 ;
	xi_u = 603 ;
	eta_v = 517 ;
variables:
	float angle(eta_rho, xi_rho) ;
		angle:_FillValue = NaNf ;
		angle:long_name = "angle between XI-axis and EAST" ;
		angle:units = "radians" ;
		angle:field = "angle, scalar" ;
	float u(time, s_rho, eta_rho, xi_u) ;
		u:_FillValue = 1.e+33f ;
		u:long_name = "averaged u-momentum component" ;
		u:units = "meter second-1" ;
		u:coordinates = "lon_u lat_u" ;
	float v(time, s_rho, eta_v, xi_rho) ;
		v:_FillValue = 1.e+33f ;
		v:long_name = "2d-slice of averaged v-momentum component" ;
		v:units = "meter second-1" ;
		v:coordinates = "lon_v lat_v" ;
	float w(time, s_rho, eta_rho, xi_rho) ;
		w:_FillValue = 1.e+33f ;
		w:long_name = "averaged vertical velocity" ;
		w:units = "meter second-1" ;
		w:coordinates = "lon_rho lat_rho" ;
	int64 time_counter(time) ;
		time_counter:units = "days since 1979-01-01T12:05:00" ;
		time_counter:calendar = "proleptic_gregorian" ;
```
The netcdf structure for depth levels:
```
netcdf depth_z_pos {
dimensions:
	eta_rho = 518 ;
	xi_rho = 604 ;
	s_rho = 64 ;
variables:
	float h(eta_rho, xi_rho) ;
		h:_FillValue = NaNf ;
		h:long_name = "bathymetry at RHO-points" ;
		h:units = "meter" ;
		h:field = "bath, scalar" ;
	double depth_Z(s_rho, eta_rho, xi_rho) ;
		depth_Z:_FillValue = NaN ;
```
I used the following code:
```
from parcels import FieldSet, ParticleSet, JITParticle, AdvectionRK4_3D, Field

data_file = "nco_time_uvw_pactcs30_1979.nc"
depthfiles = 'depth_z_pos.nc'
mesh_file = "my_grid.nc"


filenames = {'U': {'lon': mesh_file, 'lat': mesh_file, 'depth': depthfiles, 'data': data_file},
             'V': {'lon': mesh_file, 'lat': mesh_file, 'depth': depthfiles, 'data': data_file},
             'W': {'lon': mesh_file, 'lat': mesh_file, 'depth': depthfiles, 'data': data_file}}

variables = {'U': 'u',
             'V': 'v',
             'W': 'w'}

dimensions = {'lon': 'corr_lon_psi', 'lat': 'lat_psi', 'depth': 'depth_Z', 'time': 'time_counter'}
fieldset = FieldSet.from_c_grid_dataset(filenames, variables, dimensions)

New_DS = xr.open_dataset(mesh_file)
lats = New_DS.lat_psi[:,:].values
lons = New_DS.corr_lon_psi[:,:].values

fieldset.add_field(Field('bottom_depth',fieldset.U.depth[-1,:,:],lon=lons,lat=lats))
fieldset.add_field(Field('top_depth',fieldset.U.depth[0,:,:],lon=lons,lat=lats))
```
After determining random initial positions for my particles and making sure that they are in the domain and at a depth with u, v, and w I ran an adapted AdvectionRK4_3D kernel which does not allow the particles to get to the surface (to avoid getting a ThroughSurfaceError):

```
def Adapted_RK4(particle, fieldset, time):

    if particle.depth <= fieldset.top_depth[0,0,particle.lat,particle.lon]:
        particle.depth = fieldset.top_depth[0,0,particle.lat,particle.lon]+0.1
        
    (u1, v1, w1) = fieldset.UVW[time, particle.depth, particle.lat, particle.lon]
    lon1 = particle.lon + u1*.5*particle.dt
    lat1 = particle.lat + v1*.5*particle.dt
    if particle.depth + w1*.5*particle.dt <= fieldset.top_depth[0,0,lat1,lon1]:
        dep1 = fieldset.top_depth[0,0,lat1,lon1]+0.1
    else:    
        dep1 = particle.depth + w1*.5*particle.dt
        
    (u2, v2, w2) = fieldset.UVW[time + .5 * particle.dt, dep1, lat1, lon1]
    lon2 = particle.lon + u2*.5*particle.dt
    lat2 = particle.lat + v2*.5*particle.dt
    if particle.depth + w2*.5*particle.dt <= fieldset.top_depth[0,0,lat2,lon2]:
        dep2 = fieldset.top_depth[0,0,lat2,lon2]+0.1
    else:    
        dep2 = particle.depth + w2*.5*particle.dt
        
    (u3, v3, w3) = fieldset.UVW[time + .5 * particle.dt, dep2, lat2, lon2]
    lon3 = particle.lon + u3*particle.dt
    lat3 = particle.lat + v3*particle.dt
    if particle.depth + w3*particle.dt <= fieldset.top_depth[0,0,lat3,lon3]:
        dep3 = fieldset.top_depth[0,0,lat3,lon3]+0.1
    else:    
        dep3 = particle.depth + w3*particle.dt
        
    (u4, v4, w4) = fieldset.UVW[time + particle.dt, dep3, lat3, lon3]
    particle.lon += (u1 + 2*u2 + 2*u3 + u4) / 6. * particle.dt
    particle.lat += (v1 + 2*v2 + 2*v3 + v4) / 6. * particle.dt
    particle.depth += (w1 + 2*w2 + 2*w3 + w4) / 6. * particle.dt
```
Running the simulation for roughly 50 particles in Scipy mode worked perfectly and matched the results of another Lagrangian tool (ARIANE), however when changed to JITParticle it throws the error `KernelError: 0`
For the execution I used the following:
```
kernels = pset.Kernel(Adapted_RK4)
pset.execute(kernels,runtime=delta(days=30),dt=delta(days=1))
```

**Comment by erikvansebille** [comment]
Thanks for reporting @ursho2552! I'm happy to have a look what's going on

Could you copy-paste the full error message that precedes the `KernelError: 0`?

And for ease of debugging, could you share a subset of the data with me, e.g. via wetransfer or similar?

Best wishes,
Erik

**Comment by ursho2552** [comment]
Hi Erik sorry for the late reply, was out of the office for a few days.
The output preceding the `KernelError: 0` is:

```
INFO: Compiled JITParticleAdapted_RK4 ==> /tmp/parcels-43802/1fe53aa9cda6578ada4b9ccb60047d37_0.so

---------------------------------------------------------------------------
KernelError                               Traceback (most recent call last)
<ipython-input-11-c3b11f1e3e78> in <module>
      1 kernels = pset.Kernel(Adapted_RK4)
----> 2 pset.execute(kernels,runtime=delta(days=9),dt=delta(days=1))

~/anaconda3/envs/py3_parcels_original/lib/python3.8/site-packages/parcels/particleset.py in execute(self, pyfunc, endtime, runtime, dt, moviedt, recovery, output_file, movie_background_field, verbose_progress)
    473             else:
    474                 time = max(next_prelease, next_input, next_output, next_movie, endtime)
--> 475             self.kernel.execute(self, endtime=time, dt=dt, recovery=recovery, output_file=output_file)
    476             if abs(time-next_prelease) < tol:
    477                 pset_new = ParticleSet(fieldset=self.fieldset, time=time, lon=self.repeatlon,

~/anaconda3/envs/py3_parcels_original/lib/python3.8/site-packages/parcels/kernel.py in execute(self, pset, endtime, dt, recovery, output_file)
    355                     recovery_kernel = recovery_map[p.state]
    356                     p.state = ErrorCode.Success
--> 357                     recovery_kernel(p, self.fieldset, p.time)
    358 
    359             # Remove all particles that signalled deletion

~/anaconda3/envs/py3_parcels_original/lib/python3.8/site-packages/parcels/tools/error.py in recovery_kernel_error(particle, fieldset, time)
     99     """Default error kernel that throws exception"""
    100     msg = "Error: %s" % particle.exception if particle.exception else None
--> 101     raise KernelError(particle, fieldset=fieldset, msg=msg)
    102 
    103 

KernelError: 0
Particle P[47](lon=48.636505, lat=35.459423, depth=5.500549, time=0.000000)
Time: 1979-01-02T12:05:00.000000000,	timestep dt: 86400.000000
```

I checked whether the particle was out of bounds, but the following returned a velocity value:
```
n=15
fieldset.W.eval(pset[n].time,pset[n].depth,pset[n].lat,pset[n].lon,applyConversion=False)
```
For the datafiles, I restricted the time to only 10 days, hope is not too big.
Data file with u, v, w, and time_counter:
[uvw_pactcs30_10_days.nc](https://wetransfer.com/downloads/e4d78f0ae44c35ae1b3f2050dcc8b01f20191126062603/20286fecb1547c60f8718693cbd016f020191126062603/6a8e34)
Data file with the depth of the s-levels:
[depth_z_pos.nc](https://wetransfer.com/downloads/2208215622351086cb7573620582271e20191126063141/aca1e7fb7144e89e50dfb13b36f3b0c620191126063141/56cacf)
The grid description with `corr_lon_psi` and `lat_psi`:
[my_grid.nc](https://wetransfer.com/downloads/6d27a4e0577914810afa17597183ec3120191126063410/97ee1a43cb3894b16e30caa2f769879520191126063410/8adf15)

The values for the lon, lat, depth of this particle is found in:
```
fieldset.U.grid.lon[392,243]
fieldset.U.grid.lat[392,243]
DS_depth = xr.open_dataset('depth_z_pos.nc',mask_and_scale=True)
DS_depth.depth_Z[1,392,243]
```
Hope this is enough information for debugging. 
Best wishes,
Urs

**Comment by erikvansebille** [comment]
Hi Urs

I just checked your files, but the code below runs without any problems for me
```python
from parcels import FieldSet, ParticleSet, JITParticle, AdvectionRK4_3D, Field
import xarray as xr
from datetime import timedelta as delta

data_file = "uvw_pactcs30_10_days.nc"
depthfiles = 'depth_z_pos.nc'
mesh_file = "my_grid.nc"

filenames = {'U': {'lon': mesh_file, 'lat': mesh_file, 'depth': depthfiles, 'data': data_file},
             'V': {'lon': mesh_file, 'lat': mesh_file, 'depth': depthfiles, 'data': data_file},
             'W': {'lon': mesh_file, 'lat': mesh_file, 'depth': depthfiles, 'data': data_file}}

variables = {'U': 'u',
             'V': 'v',
             'W': 'w'}

dimensions = {'lon': 'corr_lon_psi', 'lat': 'lat_psi', 'depth': 'depth_Z', 'time': 'time_counter'}
fieldset = FieldSet.from_c_grid_dataset(filenames, variables, dimensions)

New_DS = xr.open_dataset(mesh_file)
lats = New_DS.lat_psi[:, :].values
lons = New_DS.corr_lon_psi[:, :].values

fieldset.add_field(Field('bottom_depth', fieldset.U.depth[-1, :, :], lon=lons, lat=lats))
fieldset.add_field(Field('top_depth', fieldset.U.depth[0, :, :], lon=lons, lat=lats))

def Adapted_RK4(particle, fieldset, time):
    if particle.depth <= fieldset.top_depth[0, 0, particle.lat, particle.lon]:
        particle.depth = fieldset.top_depth[0, 0, particle.lat, particle.lon] + 0.1

    (u1, v1, w1) = fieldset.UVW[time, particle.depth, particle.lat, particle.lon]
    lon1 = particle.lon + u1 * .5 * particle.dt
    lat1 = particle.lat + v1 * .5 * particle.dt
    if particle.depth + w1 * .5 * particle.dt <= fieldset.top_depth[0, 0, lat1, lon1]:
        dep1 = fieldset.top_depth[0, 0, lat1, lon1] + 0.1
    else:
        dep1 = particle.depth + w1 * .5 * particle.dt

    (u2, v2, w2) = fieldset.UVW[time + .5 * particle.dt, dep1, lat1, lon1]
    lon2 = particle.lon + u2 * .5 * particle.dt
    lat2 = particle.lat + v2 * .5 * particle.dt
    if particle.depth + w2 * .5 * particle.dt <= fieldset.top_depth[0, 0, lat2, lon2]:
        dep2 = fieldset.top_depth[0, 0, lat2, lon2] + 0.1
    else:
        dep2 = particle.depth + w2 * .5 * particle.dt

    (u3, v3, w3) = fieldset.UVW[time + .5 * particle.dt, dep2, lat2, lon2]
    lon3 = particle.lon + u3 * particle.dt
    lat3 = particle.lat + v3 * particle.dt
    if particle.depth + w3 * particle.dt <= fieldset.top_depth[0, 0, lat3, lon3]:
        dep3 = fieldset.top_depth[0, 0, lat3, lon3] + 0.1
    else:
        dep3 = particle.depth + w3 * particle.dt

    (u4, v4, w4) = fieldset.UVW[time + particle.dt, dep3, lat3, lon3]
    particle.lon += (u1 + 2 * u2 + 2 * u3 + u4) / 6. * particle.dt
    particle.lat += (v1 + 2 * v2 + 2 * v3 + v4) / 6. * particle.dt
    particle.depth += (w1 + 2 * w2 + 2 * w3 + w4) / 6. * particle.dt


pset = ParticleSet(fieldset, JITParticle, lon=48.636505, lat=35.459423, depth=5.500549)
kernels = pset.Kernel(Adapted_RK4)
pset.execute(kernels, runtime=delta(days=9), dt=delta(days=1))
print(pset)
```
returns `P[0](lon=48.116837, lat=36.198856, depth=7.116626, time=777600.000000)`

What version of parcels fo you have? 

And perhaps you can add a custom Recovery Kernel (either `ErrorCode.ErrorOutOfBounds` or `ErrorCode.ErrorThroughSurface`) that deals with particles near the surface? One option is to delete them (see e.g. the last line in cell 5 of [this tutorial](https://nbviewer.jupyter.org/github/OceanParcels/parcels/blob/master/parcels/examples/tutorial_Agulhasparticles.ipynb))?

**Comment by ursho2552** [comment]
Hi Erik,

thank you for the quick response. Currently I am using parcels version 2.1.2. I just ran the code you posted above and it also worked for me. However, I found that the problem might be the number of decimal places on my depth list. For instance the following works:
```
pset = ParticleSet(fieldset, JITParticle, lon=48.636505, lat=35.459423, depth=5.500549)
kernels = pset.Kernel(Adapted_RK4)
pset.execute(kernels, runtime=delta(days=9), dt=delta(days=1))
```
But using the corresponding depth as stored in depth_z_pos.nc  `depth=5.500548839569092` results in the `KernelError: 0`. Knowing this, I think I can simply use `depth=np.around(5.500548839569092,decimals=8)` to fix the problem. 

Also thank you for the Revovery Kernel hint, that will definitely be very useful!
